common/nym-lp/src/state_machine.rs

```diff
@@ -452,43 +513,108 @@ impl LpStateMachine {
+                                                     result_action = Some(Err(e));
+                                                     LpState::Closed { reason }
+                                                 }
+                                             }
+                                         }
+                                         Err(e) => {
+                                             let reason = e.to_string();
+                                             result_action = Some(Err(e));
+                                             LpState::Closed { reason }
+                                         }
+                                     }
+                                 }
+                                 Err(e) => {
+                                     let reason = e.to_string();
                                      result_action = Some(Err(e));
-                                     LpState::Transport{ session }
-                                 } else {
-                                     // 4. Deliver data
-                                     result_action = Some(Ok(LpAction::DeliverData(BytesMut::from(plaintext.as_slice()))));
-                                     // Remain in transport state
-                                     LpState::Transport { session }
+                                     LpState::Closed { reason }
                                  }
                              }
-                             Err(e) => { // Error decrypting data
-                                 let reason = e.to_string();
-                                 result_action = Some(Err(e.into()));
-                                 LpState::Closed { reason }
+                         }
+                         // Normal encrypted data
+                         LpMessage::EncryptedData(_) => {
+                             // 1. Check replay protection
+                             if let Err(e) = session.receiving_counter_quick_check(packet.header.counter) {
+                                 result_action = Some(Err(e));
+                                 LpState::Transport { session }
+                             } else {
+                                 // 2. Decrypt data
+                                 match session.decrypt_data(&packet.message) {
+                                     Ok(plaintext) => {
+                                         // 3. Mark counter as received
+                                         if let Err(e) = session.receiving_counter_mark(packet.header.counter) {
+                                             result_action = Some(Err(e));
+                                             LpState::Transport { session }
+                                         } else {
+                                             // 4. Deliver data
+                                             result_action = Some(Ok(LpAction::DeliverData(BytesMut::from(plaintext.as_slice()))));
```

Comment:
does it make sense to store the data in `BytesMut` given they should be delivered as they are, i.e. no further mutability? (ultimately they seem to be always just converted into `Vec<u8>` or dereferences into &[u8] to be further deserialised)